forecasting/src/autogluon/forecasting/models/presets.py

```diff
@@ -0,0 +1,214 @@
+                ),
+            },
+        },
+    }
+    return default_model_hps[key]
+
+
+def get_preset_models(
+    freq: str,
+    prediction_length: int,
+    path: str,
+    eval_metric: str,
+    hyperparameters: Union[str, Dict],
+    hyperparameter_tune: bool,
+    invalid_model_names: List[str],
+    **kwargs,
+):
+    """
+    Create a list of models according to hyperparameters. If hyperparamaters=None,
+    will create models according to presets.
+    """
+    models = []
+    if isinstance(hyperparameters, str):
+        hyperparameters = copy.deepcopy(
+            get_default_hps(hyperparameters, prediction_length)
+        )
+    else:
+        hp_str = "default" if not hyperparameter_tune else "default_hpo"
+        default_hps = copy.deepcopy(get_default_hps(hp_str, prediction_length))
+
+        if hyperparameters is not None:
+            # filter only default_hps for models with hyperparameters provided
+            default_hps = {
+                model: default_hps.get(model, {}) for model in hyperparameters
+            }
+            for model in hyperparameters:
+                default_hps[model].update(hyperparameters[model])
+        hyperparameters = copy.deepcopy(default_hps)
+
+    if hyperparameter_tune:
+        verify_contains_searchspace(hyperparameters)
+    else:
+        verify_no_searchspace(hyperparameters)
+
+    invalid_model_names = set(invalid_model_names)
+    all_assigned_names = set(invalid_model_names)
+
+    for model, model_hps in hyperparameters.items():
+        if isinstance(model, str):
+            if model not in MODEL_TYPES:
+                raise ValueError(f"Model {model} is not supported yet.")
+            model_type = MODEL_TYPES[model]
+        elif isinstance(model, AbstractForecastingModelFactory):
+            model_type = model
+        elif not issubclass(model, AbstractForecastingModel):
+            logger.warning(
+                f"Customized model {model} does not inherit from {AbstractForecastingModel}"
+            )
+            model_type = model
+        else:
+            logger.log(20, f"Custom Model Type Detected: {model}")
+            model_type = model
+
+        model_type_kwargs = dict(
+            path=path,
+            freq=freq,
+            prediction_length=prediction_length,
+            eval_metric=eval_metric,
+            hyperparameters=model_hps,
+            **kwargs,
+        )
+
+        # add models while preventing name collisions
+        model = model_type(**model_type_kwargs)
+        name_stem = model.name
+
+        model_type_kwargs.pop("name", None)
+        increment = 1
+        while model.name in all_assigned_names:
+            # if name collides with a name prohibited by the function argument
+            # then we always append _2
+            if model.name in invalid_model_names:
+                name_stem = model.name
+                model = model_type(name=f"{model.name}_2", **model_type_kwargs)
```

Comment:
Why? I don't see why we use this logic instead of the generic increment logic in the else statement.

It shouldn't matter if I call this function once with models A B C or three times with model A, then model B, then model C, the exact same model names should be given to A B C.

With how the code currently works, if I called once and their names were all "M", then it would be:

A.name = M
B.name = M_2
C.name = M_3

And if I called three times each with just 1 model, it would instead be:

A.name = M
B.name = M_2
C.name = M_2_2

I propose that the first result should apply always, and the entire while loop be simplified to:

```
while model.name in all_assigned_names:
    increment += 1
    model = model_type(name=f"{name_stem}_{increment}", **model_type_kwargs)
```

Reply:
I see. referring to your original comment

> I'd like to propose we use the same solution we use in TabularPredictor: appending an integer such as A and A_2. If we had 5 A models, we would then have A, A_2, A_3, A_4, A_5.
> If we have 2 A models and an A_2 model, it would then look like: A, A_2, A_2_2. This will always guarantee uniqueness, will be consistent across different fit calls, and is intuitive to the user.

We could implement this solution without `_2_2` as well. However then, the user could provide a name ending with `_\d+$` and we would disregard if such a name was provided by the user or assigned by the library in a previous round. That is, the user can have their own reasons for appending integers, such as versioning, and we would have to either 1/override these numbers to stay consistent with your ask in 

> my suggested implementation would be:
```
invalid_model_names (those from previous calls to fit): ["A", "B", "B_2"]
new default model names being pushed: ["B", "B"]
the output is ["B_3", "B_4"]
```

or 2/somehow keep track of a global state of all assigned/user-provided names also outside the scope of this function, maybe in a Trainer instance variable. I think the current version is a good tradeoff but I'm eager to hear your thoughts.

Reply:
tl; dr. In the above example, the user could be providing `B_2` as their own numbering scheme and we would be just incrementing the number to 3 regardless.

Reply:
I'm not saying that if the user provides B_2 it should become B_3, it should become B_2_2 in this case.

However, in the example, the user explicitly is passing only "B", NOT "B_2":

```
invalid_model_names (those from previous calls to fit): ["A", "B", "B_2"]
new default model names being pushed: ["B", "B"]
the output is ["B_3", "B_4"]
```

If they instead pass "B_2" (and more complex options for sake of clarity):
```
invalid_model_names (those from previous calls to fit): ["A", "B", "B_2"]
new default model names being pushed: ["B", "B_2", "B", "B_2", "B_2_3"]
the output is ["B_3", "B_2_2", "B_4", "B_2_3", "B_2_3_2"]
```

Does this make sense? We never touch the default model name, only append a suffix to it.

This is achieved via:

```
while model.name in all_assigned_names:
    increment += 1
    model = model_type(name=f"{name_stem}_{increment}", **model_type_kwargs)
```